cluster/manager/manager.go

```diff
@@ -1135,25 +1112,21 @@ func (c *ClusterManager) initListeners(
 		logrus.Infof("This node does not participates in quorum decisions")
 	}
 
-	initFunc := func(clusterInfo cluster.ClusterInfo) error {
-		numNodes := 0
-		for _, node := range clusterInfo.NodeEntries {
-			if node.Status != api.Status_STATUS_DECOMMISSION {
-				numNodes++
-			}
-		}
-		if clusterMaxSize > 0 && numNodes > clusterMaxSize {
-			// throw a slightly different error depending if node was already
-			// in the cluster.
-			if exist {
-				return ErrOldNodeOverMaxCapacity(clusterMaxSize)
+	initFunc := func(clusterInfo *cluster.ClusterInfo) error {
+		// Irrespective of whether the node is doing an Init or is
+		// already in cluster, check with listeners if it is OK to join
+		// this cluster.
+		for e := c.listeners.Front(); e != nil; e = e.Next() {
```

Comment:
there is already a finalizeCB callback, so not sure why we want to callback listeners again ?

Reply:
We want the clusterMaxSize check to be called during Init as well as Join. This piece of code (initFunc) gets called during both. While in Init finalizeCbs is populated, but in Join it is an empty array.

I have added a comment below about finalizeCbs being nil